source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,423 @@
+ * By using the inline handle, the filter state could get the key/value pair without any hash.
+ *
+ * This class also provides the interface to access the key/value pair by the normal key and the
+ * interface has similar searching performance as the normal hash map. But the insertion and removal
+ * by the normal key is slower than the normal hash map.
+ *
+ *
+ * These is usage example:
+ *
+ *   // Define registry to store the inline keys. The registry should have lifetime that no shorter
+ *   // than all inline maps created by the registry.
+ *   InlineMapRegistry registry;
+ *
+ *   // Register the inline key. We should never do this after bootstrapping. Typically, we can
+ *   // do this by define a global variable.
+ *   InlineMapRegistry<std::string>::InlineHandle inline_handle =
+ *     registry.registerInlineKey("inline_key");
+ *
+ *   // Finalize the registry. No further changes are allowed to the registry after this point.
+ *   registry.finalize();
+ *
+ *   // Create the inline map.
+ *   InlineMapPtr inline_map = registry.createInlineMap<std::string>();
+ *
+ *   // Get value by inline handle.
+ *   inline_map->insert(inline_handle, std::make_unique<std::string>("value"));
+ *   EXPECT_EQ(*inline_map->lookup(inline_handle), "value");
+ */
+template <class Key> class InlineMapRegistry {
+public:
+  /**
+   * This is the handle used to access the key/value pair in the inline map. The handle is
+   * lightweight and could be copied and passed by value.
+   */
+  class InlineKey {
+  public:
+    /**
+     * Get the id of the inline entry in the inline array. This could be used to access the
+     * key/value pair in the inline map without hash searching.
+     */
+    uint64_t inlineId() const { return inline_id_; }
+
+  private:
+    friend class InlineMapRegistry;
+
+    // This constructor should only be called by InlineMapRegistry.
+    InlineKey(uint64_t inline_id) : inline_id_(inline_id) {}
+
+    uint64_t inline_id_{};
+
+#ifndef NDEBUG
+  public:
+    // This is the registry id that the inline key created by. This is used to validate the
+    // inline key validity.
+    uint64_t registryId() const { return registry_id_; }
+
+  private:
+    InlineKey(uint64_t inline_id, uint64_t registry_id)
+        : inline_id_(inline_id), registry_id_(registry_id) {}
+    uint64_t registry_id_{};
+#endif
+  };
+
+  using RegistrationMap = absl::flat_hash_map<Key, uint64_t>;
+  using RegistrationSet = std::vector<Key>;
+
+  template <class T> class InlineMap : public InlineStorage {
+  public:
+    using TPtr = std::unique_ptr<T>;
+    using UnderlayHashMap = absl::flat_hash_map<Key, TPtr>;
+
+    // Get the entry for the given key. If the key is not found, return nullptr.
+    template <class ArgK> const T* lookup(const ArgK& key) const { return lookupImpl(key); }
+    template <class ArgK> T* lookup(const ArgK& key) { return lookupImpl(key); }
```

Comment:
My first instinct is to try to mimic STL map's interface, but I realize adding an STL-compliant iterator may be onerous.

But I think you could provider `operator[]` which is very useful for cases where you don't care if you are inserting or overwrite, and a variant of insert() that can return (say) an `OptRef<Value>` or similar, so you can do something slightly different between a new value and an overwrite, like I suggested for your implementation for creating a new inline key.

Reply:
operator[] is supported and now insert will return false and fail if a key is exist.